Apollo compiler failing for valid query

Apollo compiler generates code for the following query:
 
```
fragment Setting on Setting {
  __typename
  value {
    __typename
    ... on StringListSettingValue {
      list
    }
  }
}

query queryName($key: String!, $nodeID: String!, $subkey: String) {
  queryName(key: $key, nodeID: $nodeID, subkey: $subkey) {
    ...Setting
  }
}
```
But starts failing for the following valid query:

```
fragment Setting on Setting {
  __typename
  value {
    __typename
    ... on StringListSettingValue {
      list
    }
  }
  ... on SelectSetting {
    options {
      allowFreeText
      id
      label
    }
  }
}

query queryName($key: String!, $nodeID: String!, $subkey: String) {
  queryName(key: $key, nodeID: $nodeID, subkey: $subkey) {
    ...Setting
  }
}

```
with this error:

```
* What went wrong:
Execution failed for task ':domain:generateDebugApolloClasses'.
> non-abstract type AsStringListSettingValue cannot declare abstract method __typename

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':domain:generateDebugApolloClasses'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipCachedTaskExecuter.execute(SkipCachedTaskExecuter.java:108)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ResolveBuildCacheKeyExecuter.execute(ResolveBuildCacheKeyExecuter.java:61)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:60)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:626)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:581)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.IllegalArgumentException: non-abstract type AsStringListSettingValue cannot declare abstract method __typename
        at com.squareup.javapoet.Util.checkArgument(Util.java:64)
        at com.squareup.javapoet.TypeSpec$Builder.build(TypeSpec.java:585)
        at com.apollographql.apollo.compiler.UtilKt.confirmToProtocol(Util.kt:442)
        at com.apollographql.apollo.compiler.SchemaTypeSpecBuilder.interfaceTypeSpec(SchemaTypeSpecBuilder.kt:61)
        at com.apollographql.apollo.compiler.SchemaTypeSpecBuilder.build(SchemaTypeSpecBuilder.kt:27)
        at com.apollographql.apollo.compiler.ir.Field.toTypeSpec(Field.kt:35)
        at com.apollographql.apollo.compiler.SchemaTypeSpecBuilder.nestedTypeSpecs(SchemaTypeSpecBuilder.kt:110)
        at com.apollographql.apollo.compiler.SchemaTypeSpecBuilder.interfaceTypeSpec(SchemaTypeSpecBuilder.kt:34)
        at com.apollographql.apollo.compiler.SchemaTypeSpecBuilder.build(SchemaTypeSpecBuilder.kt:27)
        at com.apollographql.apollo.compiler.ir.Fragment.toTypeSpec(Fragment.kt:32)
        at com.apollographql.apollo.compiler.ir.CodeGenerator$DefaultImpls.toTypeSpec$default(CodeGenerator.kt:6)
        at com.apollographql.apollo.compiler.GraphQLCompiler.writeJavaFiles(GraphQLCompiler.kt:42)
        at com.apollographql.apollo.compiler.GraphQLCompiler.write(GraphQLCompiler.kt:32)
        at com.apollographql.apollo.gradle.ApolloClassGenTask$1.execute(ApolloClassGenTask.java:45)
        at com.apollographql.apollo.gradle.ApolloClassGenTask$1.execute(ApolloClassGenTask.java:39)
        at org.gradle.api.internal.changedetection.changes.RebuildIncrementalTaskInputs.doOutOfDate(RebuildIncrementalTaskInputs.java:43)
        at org.gradle.api.internal.changedetection.changes.StatefulIncrementalTaskInputs.outOfDate(StatefulIncrementalTaskInputs.java:40)
        at org.gradle.api.internal.changedetection.changes.RebuildIncrementalTaskInputs.outOfDate(RebuildIncrementalTaskInputs.java:27)
        at com.apollographql.apollo.gradle.ApolloClassGenTask.generateClasses(ApolloClassGenTask.java:39)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.IncrementalTaskAction.doExecute(IncrementalTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:121)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        ... 31 more



```

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/SchemaTypeSpecBuilder.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/SchemaTypeSpecBuilder.kt
@@ -115,7 +115,7 @@ class SchemaTypeSpecBuilder(
   }
 
   private fun inlineFragmentsTypeSpecs(): List<Pair<String, TypeSpec>> =
-      inlineFragments.map { it.formatClassName() to it.toTypeSpec(context = context, abstract = abstract) }
+      inlineFragments.map { it.formatClassName() to it.toTypeSpec(context = context, abstract = false) }
 
   private fun fragmentsAccessorMethodSpec(): MethodSpec {
     val fragmentsName = FRAGMENTS_FIELD.name

--- a/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt
+++ b/apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt
@@ -442,21 +442,23 @@ fun TypeSpec.confirmToProtocol(protocolSpec: TypeSpec): TypeSpec {
         }
       }.filter { (_, nestedProtocol) -> nestedProtocol != null }
 
-  return TypeSpec.classBuilder(name)
-      .superclass(superclass)
+  val classBuilder = (kind != TypeSpec.Kind.INTERFACE)
+  val typeSpec = if (classBuilder) TypeSpec.classBuilder(name) else TypeSpec.interfaceBuilder(name)
+  return typeSpec
+      .also { if (classBuilder) it.superclass(superclass) }
       .addJavadoc(javadoc)
       .addAnnotations(annotations)
       .addModifiers(*modifiers.toTypedArray())
       .addSuperinterfaces(superinterfaces)
       .addSuperinterface(ClassName.get("", protocolSpec.name))
-      .addFields(fieldSpecs)
+      .also { if (classBuilder) it.addFields(fieldSpecs) }
+      .addMethods(methodSpecs)
       .addTypes(typeSpecs.map { typeSpec ->
         val protocol = nestedTypeProtocols[typeSpec.name]
         protocol?.let { typeSpec.confirmToProtocol(it) } ?: typeSpec
       })
-      .addMethods(methodSpecs)
-      .let { if (initializerBlock.isEmpty) it else it.addInitializerBlock(initializerBlock) }
-      .let { if (staticBlock.isEmpty) it else it.addStaticBlock(staticBlock) }
+      .also { if (!staticBlock.isEmpty) it.addStaticBlock(staticBlock) }
+      .also { if (classBuilder && !initializerBlock.isEmpty) it.addInitializerBlock(initializerBlock) }
       .build()
 }
 

--- a/apollo-compiler/src/test/graphql/com/example/nested_inline_fragment/TestOperation.json
+++ b/apollo-compiler/src/test/graphql/com/example/nested_inline_fragment/TestOperation.json
@@ -1,0 +1,210 @@
+{
+  "operations": [
+    {
+      "filePath": "src/test/graphql/com/example/nested_inline_fragment/TestOperation.graphql",
+      "operationName": "TestQuery",
+      "operationType": "query",
+      "rootType": "Query",
+      "variables": [],
+      "source": "query TestOperation {\n  setting {\n    __typename\n    ...TestSetting\n  }\n}",
+      "fields": [
+        {
+          "responseName": "setting",
+          "fieldName": "setting",
+          "type": "Setting!",
+          "args": [],
+          "isConditional": false,
+          "isDeprecated": false,
+          "deprecationReason": "",
+          "fields": [
+            {
+              "responseName": "__typename",
+              "fieldName": "__typename",
+              "type": "String!",
+              "isConditional": false
+            }
+          ],
+          "fragmentSpreads": [
+            "TestSetting"
+          ],
+          "inlineFragments": []
+        }
+      ],
+      "fragmentSpreads": [],
+      "inlineFragments": [],
+      "fragmentsReferenced": [
+        "TestSetting"
+      ],
+      "sourceWithFragments": "query TestOperation {\n  setting {\n    __typename\n    ...TestSetting\n  }\n}\nfragment TestSetting on Setting {\n  __typename\n  value {\n    __typename\n    ... on StringListSettingValue {\n      list\n    }\n  }\n  ... on SelectSetting {\n    options {\n      __typename\n      allowFreeText\n      id\n      label\n    }\n  }\n}",
+      "operationId": "c4e875b8b3292e1ca6a36e8dccd11c724ee40eed3a3b87a1107fceddb3186fd2"
+    }
+  ],
+  "fragments": [
+    {
+      "typeCondition": "Setting",
+      "possibleTypes": [
+        "BooleanSetting",
+        "SelectSetting",
+        "StringListSetting",
+        "TextSetting",
+        "TypedStringListSetting"
+      ],
+      "fragmentName": "TestSetting",
+      "filePath": "src/test/graphql/com/example/nested_inline_fragment/TestOperation.graphql",
+      "source": "fragment TestSetting on Setting {\n  __typename\n  value {\n    __typename\n    ... on StringListSettingValue {\n      list\n    }\n  }\n  ... on SelectSetting {\n    options {\n      __typename\n      allowFreeText\n      id\n      label\n    }\n  }\n}",
+      "fields": [
+        {
+          "responseName": "__typename",
+          "fieldName": "__typename",
+          "type": "String!",
+          "isConditional": false
+        },
+        {
+          "responseName": "value",
+          "fieldName": "value",
+          "type": "SettingValue!",
+          "isConditional": false,
+          "isDeprecated": false,
+          "deprecationReason": "",
+          "fields": [
+            {
+              "responseName": "__typename",
+              "fieldName": "__typename",
+              "type": "String!",
+              "isConditional": false
+            }
+          ],
+          "fragmentSpreads": [],
+          "inlineFragments": [
+            {
+              "typeCondition": "StringListSettingValue",
+              "possibleTypes": [
+                "StringListSettingValue"
+              ],
+              "fields": [
+                {
+                  "responseName": "__typename",
+                  "fieldName": "__typename",
+                  "type": "String!",
+                  "isConditional": false
+                },
+                {
+                  "responseName": "list",
+                  "fieldName": "list",
+                  "type": "[String!]",
+                  "isConditional": false,
+                  "isDeprecated": false,
+                  "deprecationReason": ""
+                }
+              ],
+              "fragmentSpreads": []
+            }
+          ]
+        }
+      ],
+      "fragmentSpreads": [],
+      "inlineFragments": [
+        {
+          "typeCondition": "SelectSetting",
+          "possibleTypes": [
+            "SelectSetting"
+          ],
+          "fields": [
+            {
+              "responseName": "__typename",
+              "fieldName": "__typename",
+              "type": "String!",
+              "isConditional": false
+            },
+            {
+              "responseName": "value",
+              "fieldName": "value",
+              "type": "SettingValue!",
+              "isConditional": false,
+              "isDeprecated": false,
+              "deprecationReason": "",
+              "fields": [
+                {
+                  "responseName": "__typename",
+                  "fieldName": "__typename",
+                  "type": "String!",
+                  "isConditional": false
+                }
+              ],
+              "fragmentSpreads": [],
+              "inlineFragments": [
+                {
+                  "typeCondition": "StringListSettingValue",
+                  "possibleTypes": [
+                    "StringListSettingValue"
+                  ],
+                  "fields": [
+                    {
+                      "responseName": "__typename",
+                      "fieldName": "__typename",
+                      "type": "String!",
+                      "isConditional": false
+                    },
+                    {
+                      "responseName": "list",
+                      "fieldName": "list",
+                      "type": "[String!]",
+                      "isConditional": false,
+                      "isDeprecated": false,
+                      "deprecationReason": ""
+                    }
+                  ],
+                  "fragmentSpreads": []
+                }
+              ]
+            },
+            {
+              "responseName": "options",
+              "fieldName": "options",
+              "type": "[SelectableItem!]",
+              "isConditional": false,
+              "isDeprecated": false,
+              "deprecationReason": "",
+              "fields": [
+                {
+                  "responseName": "__typename",
+                  "fieldName": "__typename",
+                  "type": "String!",
+                  "isConditional": false
+                },
+                {
+                  "responseName": "allowFreeText",
+                  "fieldName": "allowFreeText",
+                  "type": "Boolean!",
+                  "isConditional": false,
+                  "isDeprecated": false,
+                  "deprecationReason": ""
+                },
+                {
+                  "responseName": "id",
+                  "fieldName": "id",
+                  "type": "String!",
+                  "isConditional": false,
+                  "isDeprecated": false,
+                  "deprecationReason": ""
+                },
+                {
+                  "responseName": "label",
+                  "fieldName": "label",
+                  "type": "String!",
+                  "isConditional": false,
+                  "isDeprecated": false,
+                  "deprecationReason": ""
+                }
+              ],
+              "fragmentSpreads": [],
+              "inlineFragments": []
+            }
+          ],
+          "fragmentSpreads": []
+        }
+      ]
+    }
+  ],
+  "typesUsed": []
+}

--- a/apollo-compiler/src/test/graphql/com/example/nested_inline_fragment/TestQuery.java
+++ b/apollo-compiler/src/test/graphql/com/example/nested_inline_fragment/TestQuery.java
@@ -1,0 +1,347 @@
+package com.example.nested_inline_fragment;
+
+import com.apollographql.apollo.api.FragmentResponseFieldMapper;
+import com.apollographql.apollo.api.Operation;
+import com.apollographql.apollo.api.OperationName;
+import com.apollographql.apollo.api.Query;
+import com.apollographql.apollo.api.ResponseField;
+import com.apollographql.apollo.api.ResponseFieldMapper;
+import com.apollographql.apollo.api.ResponseFieldMarshaller;
+import com.apollographql.apollo.api.ResponseReader;
+import com.apollographql.apollo.api.ResponseWriter;
+import com.apollographql.apollo.api.internal.Optional;
+import com.apollographql.apollo.api.internal.Utils;
+import com.example.nested_inline_fragment.fragment.TestSetting;
+import java.lang.Object;
+import java.lang.Override;
+import java.lang.String;
+import java.util.Arrays;
+import java.util.Collections;
+import javax.annotation.Generated;
+import javax.annotation.Nonnull;
+
+@Generated("Apollo GraphQL")
+public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery.Data>, Operation.Variables> {
+  public static final String OPERATION_DEFINITION = "query TestOperation {\n"
+      + "  setting {\n"
+      + "    __typename\n"
+      + "    ...TestSetting\n"
+      + "  }\n"
+      + "}";
+
+  public static final String OPERATION_ID = "c4e875b8b3292e1ca6a36e8dccd11c724ee40eed3a3b87a1107fceddb3186fd2";
+
+  public static final String QUERY_DOCUMENT = OPERATION_DEFINITION + "\n"
+   + TestSetting.FRAGMENT_DEFINITION;
+
+  public static final OperationName OPERATION_NAME = new OperationName() {
+    @Override
+    public String name() {
+      return "TestQuery";
+    }
+  };
+
+  private final Operation.Variables variables;
+
+  public TestQuery() {
+    this.variables = Operation.EMPTY_VARIABLES;
+  }
+
+  @Override
+  public String operationId() {
+    return OPERATION_ID;
+  }
+
+  @Override
+  public String queryDocument() {
+    return QUERY_DOCUMENT;
+  }
+
+  @Override
+  public Optional<TestQuery.Data> wrapData(TestQuery.Data data) {
+    return Optional.fromNullable(data);
+  }
+
+  @Override
+  public Operation.Variables variables() {
+    return variables;
+  }
+
+  @Override
+  public ResponseFieldMapper<TestQuery.Data> responseFieldMapper() {
+    return new Data.Mapper();
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  @Override
+  public OperationName name() {
+    return OPERATION_NAME;
+  }
+
+  public static final class Builder {
+    Builder() {
+    }
+
+    public TestQuery build() {
+      return new TestQuery();
+    }
+  }
+
+  public static class Data implements Operation.Data {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forObject("setting", "setting", null, false, Collections.<ResponseField.Condition>emptyList())
+    };
+
+    final @Nonnull Setting setting;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public Data(@Nonnull Setting setting) {
+      this.setting = Utils.checkNotNull(setting, "setting == null");
+    }
+
+    public @Nonnull Setting setting() {
+      return this.setting;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeObject($responseFields[0], setting.marshaller());
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "Data{"
+          + "setting=" + setting
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof Data) {
+        Data that = (Data) o;
+        return this.setting.equals(that.setting);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= setting.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<Data> {
+      final Setting.Mapper settingFieldMapper = new Setting.Mapper();
+
+      @Override
+      public Data map(ResponseReader reader) {
+        final Setting setting = reader.readObject($responseFields[0], new ResponseReader.ObjectReader<Setting>() {
+          @Override
+          public Setting read(ResponseReader reader) {
+            return settingFieldMapper.map(reader);
+          }
+        });
+        return new Data(setting);
+      }
+    }
+  }
+
+  public static class Setting {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forString("__typename", "__typename", null, false, Collections.<ResponseField.Condition>emptyList()),
+      ResponseField.forFragment("__typename", "__typename", Arrays.asList("BooleanSetting",
+      "SelectSetting",
+      "StringListSetting",
+      "TextSetting",
+      "TypedStringListSetting"))
+    };
+
+    final @Nonnull String __typename;
+
+    private final @Nonnull Fragments fragments;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public Setting(@Nonnull String __typename, @Nonnull Fragments fragments) {
+      this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+      this.fragments = Utils.checkNotNull(fragments, "fragments == null");
+    }
+
+    public @Nonnull String __typename() {
+      return this.__typename;
+    }
+
+    public @Nonnull Fragments fragments() {
+      return this.fragments;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeString($responseFields[0], __typename);
+          fragments.marshaller().marshal(writer);
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "Setting{"
+          + "__typename=" + __typename + ", "
+          + "fragments=" + fragments
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof Setting) {
+        Setting that = (Setting) o;
+        return this.__typename.equals(that.__typename)
+         && this.fragments.equals(that.fragments);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= __typename.hashCode();
+        h *= 1000003;
+        h ^= fragments.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static class Fragments {
+      final @Nonnull TestSetting testSetting;
+
+      private volatile String $toString;
+
+      private volatile int $hashCode;
+
+      private volatile boolean $hashCodeMemoized;
+
+      public Fragments(@Nonnull TestSetting testSetting) {
+        this.testSetting = Utils.checkNotNull(testSetting, "testSetting == null");
+      }
+
+      public @Nonnull TestSetting testSetting() {
+        return this.testSetting;
+      }
+
+      public ResponseFieldMarshaller marshaller() {
+        return new ResponseFieldMarshaller() {
+          @Override
+          public void marshal(ResponseWriter writer) {
+            final TestSetting $testSetting = testSetting;
+            if ($testSetting != null) {
+              $testSetting.marshaller().marshal(writer);
+            }
+          }
+        };
+      }
+
+      @Override
+      public String toString() {
+        if ($toString == null) {
+          $toString = "Fragments{"
+            + "testSetting=" + testSetting
+            + "}";
+        }
+        return $toString;
+      }
+
+      @Override
+      public boolean equals(Object o) {
+        if (o == this) {
+          return true;
+        }
+        if (o instanceof Fragments) {
+          Fragments that = (Fragments) o;
+          return this.testSetting.equals(that.testSetting);
+        }
+        return false;
+      }
+
+      @Override
+      public int hashCode() {
+        if (!$hashCodeMemoized) {
+          int h = 1;
+          h *= 1000003;
+          h ^= testSetting.hashCode();
+          $hashCode = h;
+          $hashCodeMemoized = true;
+        }
+        return $hashCode;
+      }
+
+      public static final class Mapper implements FragmentResponseFieldMapper<Fragments> {
+        final TestSetting.Mapper testSettingFieldMapper = new TestSetting.Mapper();
+
+        @Override
+        public @Nonnull Fragments map(ResponseReader reader, @Nonnull String conditionalType) {
+          TestSetting testSetting = null;
+          if (TestSetting.POSSIBLE_TYPES.contains(conditionalType)) {
+            testSetting = testSettingFieldMapper.map(reader);
+          }
+          return new Fragments(Utils.checkNotNull(testSetting, "testSetting == null"));
+        }
+      }
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<Setting> {
+      final Fragments.Mapper fragmentsFieldMapper = new Fragments.Mapper();
+
+      @Override
+      public Setting map(ResponseReader reader) {
+        final String __typename = reader.readString($responseFields[0]);
+        final Fragments fragments = reader.readConditional($responseFields[1], new ResponseReader.ConditionalTypeReader<Fragments>() {
+          @Override
+          public Fragments read(String conditionalType, ResponseReader reader) {
+            return fragmentsFieldMapper.map(reader, conditionalType);
+          }
+        });
+        return new Setting(__typename, fragments);
+      }
+    }
+  }
+}

--- a/apollo-compiler/src/test/graphql/com/example/nested_inline_fragment/fragment/TestSetting.java
+++ b/apollo-compiler/src/test/graphql/com/example/nested_inline_fragment/fragment/TestSetting.java
@@ -1,0 +1,550 @@
+package com.example.nested_inline_fragment.fragment;
+
+import com.apollographql.apollo.api.GraphqlFragment;
+import com.apollographql.apollo.api.ResponseField;
+import com.apollographql.apollo.api.ResponseFieldMapper;
+import com.apollographql.apollo.api.ResponseFieldMarshaller;
+import com.apollographql.apollo.api.ResponseReader;
+import com.apollographql.apollo.api.ResponseWriter;
+import com.apollographql.apollo.api.internal.Optional;
+import com.apollographql.apollo.api.internal.Utils;
+import java.lang.Object;
+import java.lang.Override;
+import java.lang.String;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Generated;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+@Generated("Apollo GraphQL")
+public interface TestSetting extends GraphqlFragment {
+  String FRAGMENT_DEFINITION = "fragment TestSetting on Setting {\n"
+      + "  __typename\n"
+      + "  value {\n"
+      + "    __typename\n"
+      + "    ... on StringListSettingValue {\n"
+      + "      list\n"
+      + "    }\n"
+      + "  }\n"
+      + "  ... on SelectSetting {\n"
+      + "    options {\n"
+      + "      __typename\n"
+      + "      allowFreeText\n"
+      + "      id\n"
+      + "      label\n"
+      + "    }\n"
+      + "  }\n"
+      + "}";
+
+  List<String> POSSIBLE_TYPES = Collections.unmodifiableList(Arrays.asList( "BooleanSetting", "SelectSetting", "StringListSetting", "TextSetting", "TypedStringListSetting"));
+
+  @Nonnull String __typename();
+
+  Optional<? extends Value> value();
+
+  ResponseFieldMarshaller marshaller();
+
+  final class Mapper implements ResponseFieldMapper<TestSetting> {
+    final AsSelectSetting.Mapper asSelectSettingFieldMapper = new AsSelectSetting.Mapper();
+
+    @Override
+    public TestSetting map(ResponseReader reader) {
+      final AsSelectSetting asSelectSetting = reader.readConditional(ResponseField.forInlineFragment("__typename", "__typename", Arrays.asList("SelectSetting")), new ResponseReader.ConditionalTypeReader<AsSelectSetting>() {
+        @Override
+        public AsSelectSetting read(String conditionalType, ResponseReader reader) {
+          return asSelectSettingFieldMapper.map(reader);
+        }
+      });
+      if (asSelectSetting != null) {
+        return asSelectSetting;
+      }
+      return null;
+    }
+  }
+
+  interface Value {
+    @Nonnull String __typename();
+
+    ResponseFieldMarshaller marshaller();
+
+    final class Mapper implements ResponseFieldMapper<Value> {
+      final AsStringListSettingValue.Mapper asStringListSettingValueFieldMapper = new AsStringListSettingValue.Mapper();
+
+      @Override
+      public Value map(ResponseReader reader) {
+        final AsStringListSettingValue asStringListSettingValue = reader.readConditional(ResponseField.forInlineFragment("__typename", "__typename", Arrays.asList("StringListSettingValue")), new ResponseReader.ConditionalTypeReader<AsStringListSettingValue>() {
+          @Override
+          public AsStringListSettingValue read(String conditionalType, ResponseReader reader) {
+            return asStringListSettingValueFieldMapper.map(reader);
+          }
+        });
+        if (asStringListSettingValue != null) {
+          return asStringListSettingValue;
+        }
+        return null;
+      }
+    }
+  }
+
+  class AsStringListSettingValue implements Value {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forString("__typename", "__typename", null, false, Collections.<ResponseField.Condition>emptyList()),
+      ResponseField.forList("list", "list", null, true, Collections.<ResponseField.Condition>emptyList())
+    };
+
+    final @Nonnull String __typename;
+
+    final Optional<List<String>> list;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public AsStringListSettingValue(@Nonnull String __typename, @Nullable List<String> list) {
+      this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+      this.list = Optional.fromNullable(list);
+    }
+
+    public @Nonnull String __typename() {
+      return this.__typename;
+    }
+
+    public Optional<List<String>> list() {
+      return this.list;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeString($responseFields[0], __typename);
+          writer.writeList($responseFields[1], list.isPresent() ? list.get() : null, new ResponseWriter.ListWriter() {
+            @Override
+            public void write(Object value, ResponseWriter.ListItemWriter listItemWriter) {
+              listItemWriter.writeString(value);
+            }
+          });
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "AsStringListSettingValue{"
+          + "__typename=" + __typename + ", "
+          + "list=" + list
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof AsStringListSettingValue) {
+        AsStringListSettingValue that = (AsStringListSettingValue) o;
+        return this.__typename.equals(that.__typename)
+         && this.list.equals(that.list);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= __typename.hashCode();
+        h *= 1000003;
+        h ^= list.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<AsStringListSettingValue> {
+      @Override
+      public AsStringListSettingValue map(ResponseReader reader) {
+        final String __typename = reader.readString($responseFields[0]);
+        final List<String> list = reader.readList($responseFields[1], new ResponseReader.ListReader<String>() {
+          @Override
+          public String read(ResponseReader.ListItemReader listItemReader) {
+            return listItemReader.readString();
+          }
+        });
+        return new AsStringListSettingValue(__typename, list);
+      }
+    }
+  }
+
+  class AsSelectSetting implements TestSetting {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forString("__typename", "__typename", null, false, Collections.<ResponseField.Condition>emptyList()),
+      ResponseField.forObject("value", "value", null, true, Collections.<ResponseField.Condition>emptyList()),
+      ResponseField.forList("options", "options", null, true, Collections.<ResponseField.Condition>emptyList())
+    };
+
+    final @Nonnull String __typename;
+
+    final Optional<Value1> value;
+
+    final Optional<List<Option>> options;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public AsSelectSetting(@Nonnull String __typename, @Nullable Value1 value,
+        @Nullable List<Option> options) {
+      this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+      this.value = Optional.fromNullable(value);
+      this.options = Optional.fromNullable(options);
+    }
+
+    public @Nonnull String __typename() {
+      return this.__typename;
+    }
+
+    public Optional<Value1> value() {
+      return this.value;
+    }
+
+    public Optional<List<Option>> options() {
+      return this.options;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeString($responseFields[0], __typename);
+          writer.writeObject($responseFields[1], value.isPresent() ? value.get().marshaller() : null);
+          writer.writeList($responseFields[2], options.isPresent() ? options.get() : null, new ResponseWriter.ListWriter() {
+            @Override
+            public void write(Object value, ResponseWriter.ListItemWriter listItemWriter) {
+              listItemWriter.writeObject(((Option) value).marshaller());
+            }
+          });
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "AsSelectSetting{"
+          + "__typename=" + __typename + ", "
+          + "value=" + value + ", "
+          + "options=" + options
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof AsSelectSetting) {
+        AsSelectSetting that = (AsSelectSetting) o;
+        return this.__typename.equals(that.__typename)
+         && this.value.equals(that.value)
+         && this.options.equals(that.options);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= __typename.hashCode();
+        h *= 1000003;
+        h ^= value.hashCode();
+        h *= 1000003;
+        h ^= options.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<AsSelectSetting> {
+      final Value1.Mapper value1FieldMapper = new Value1.Mapper();
+
+      final Option.Mapper optionFieldMapper = new Option.Mapper();
+
+      @Override
+      public AsSelectSetting map(ResponseReader reader) {
+        final String __typename = reader.readString($responseFields[0]);
+        final Value1 value = reader.readObject($responseFields[1], new ResponseReader.ObjectReader<Value1>() {
+          @Override
+          public Value1 read(ResponseReader reader) {
+            return value1FieldMapper.map(reader);
+          }
+        });
+        final List<Option> options = reader.readList($responseFields[2], new ResponseReader.ListReader<Option>() {
+          @Override
+          public Option read(ResponseReader.ListItemReader listItemReader) {
+            return listItemReader.readObject(new ResponseReader.ObjectReader<Option>() {
+              @Override
+              public Option read(ResponseReader reader) {
+                return optionFieldMapper.map(reader);
+              }
+            });
+          }
+        });
+        return new AsSelectSetting(__typename, value, options);
+      }
+    }
+  }
+
+  interface Value1 extends Value {
+    @Nonnull String __typename();
+
+    ResponseFieldMarshaller marshaller();
+
+    final class Mapper implements ResponseFieldMapper<Value1> {
+      final AsStringListSettingValue1.Mapper asStringListSettingValue1FieldMapper = new AsStringListSettingValue1.Mapper();
+
+      @Override
+      public Value1 map(ResponseReader reader) {
+        final AsStringListSettingValue1 asStringListSettingValue = reader.readConditional(ResponseField.forInlineFragment("__typename", "__typename", Arrays.asList("StringListSettingValue")), new ResponseReader.ConditionalTypeReader<AsStringListSettingValue1>() {
+          @Override
+          public AsStringListSettingValue1 read(String conditionalType, ResponseReader reader) {
+            return asStringListSettingValue1FieldMapper.map(reader);
+          }
+        });
+        if (asStringListSettingValue != null) {
+          return asStringListSettingValue;
+        }
+        return null;
+      }
+    }
+  }
+
+  class AsStringListSettingValue1 implements Value1 {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forString("__typename", "__typename", null, false, Collections.<ResponseField.Condition>emptyList()),
+      ResponseField.forList("list", "list", null, true, Collections.<ResponseField.Condition>emptyList())
+    };
+
+    final @Nonnull String __typename;
+
+    final Optional<List<String>> list;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public AsStringListSettingValue1(@Nonnull String __typename, @Nullable List<String> list) {
+      this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+      this.list = Optional.fromNullable(list);
+    }
+
+    public @Nonnull String __typename() {
+      return this.__typename;
+    }
+
+    public Optional<List<String>> list() {
+      return this.list;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeString($responseFields[0], __typename);
+          writer.writeList($responseFields[1], list.isPresent() ? list.get() : null, new ResponseWriter.ListWriter() {
+            @Override
+            public void write(Object value, ResponseWriter.ListItemWriter listItemWriter) {
+              listItemWriter.writeString(value);
+            }
+          });
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "AsStringListSettingValue1{"
+          + "__typename=" + __typename + ", "
+          + "list=" + list
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof AsStringListSettingValue1) {
+        AsStringListSettingValue1 that = (AsStringListSettingValue1) o;
+        return this.__typename.equals(that.__typename)
+         && this.list.equals(that.list);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= __typename.hashCode();
+        h *= 1000003;
+        h ^= list.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<AsStringListSettingValue1> {
+      @Override
+      public AsStringListSettingValue1 map(ResponseReader reader) {
+        final String __typename = reader.readString($responseFields[0]);
+        final List<String> list = reader.readList($responseFields[1], new ResponseReader.ListReader<String>() {
+          @Override
+          public String read(ResponseReader.ListItemReader listItemReader) {
+            return listItemReader.readString();
+          }
+        });
+        return new AsStringListSettingValue1(__typename, list);
+      }
+    }
+  }
+
+  class Option {
+    static final ResponseField[] $responseFields = {
+      ResponseField.forString("__typename", "__typename", null, false, Collections.<ResponseField.Condition>emptyList()),
+      ResponseField.forBoolean("allowFreeText", "allowFreeText", null, false, Collections.<ResponseField.Condition>emptyList()),
+      ResponseField.forString("id", "id", null, false, Collections.<ResponseField.Condition>emptyList()),
+      ResponseField.forString("label", "label", null, false, Collections.<ResponseField.Condition>emptyList())
+    };
+
+    final @Nonnull String __typename;
+
+    final boolean allowFreeText;
+
+    final @Nonnull String id;
+
+    final @Nonnull String label;
+
+    private volatile String $toString;
+
+    private volatile int $hashCode;
+
+    private volatile boolean $hashCodeMemoized;
+
+    public Option(@Nonnull String __typename, boolean allowFreeText, @Nonnull String id,
+        @Nonnull String label) {
+      this.__typename = Utils.checkNotNull(__typename, "__typename == null");
+      this.allowFreeText = allowFreeText;
+      this.id = Utils.checkNotNull(id, "id == null");
+      this.label = Utils.checkNotNull(label, "label == null");
+    }
+
+    public @Nonnull String __typename() {
+      return this.__typename;
+    }
+
+    public boolean allowFreeText() {
+      return this.allowFreeText;
+    }
+
+    public @Nonnull String id() {
+      return this.id;
+    }
+
+    public @Nonnull String label() {
+      return this.label;
+    }
+
+    public ResponseFieldMarshaller marshaller() {
+      return new ResponseFieldMarshaller() {
+        @Override
+        public void marshal(ResponseWriter writer) {
+          writer.writeString($responseFields[0], __typename);
+          writer.writeBoolean($responseFields[1], allowFreeText);
+          writer.writeString($responseFields[2], id);
+          writer.writeString($responseFields[3], label);
+        }
+      };
+    }
+
+    @Override
+    public String toString() {
+      if ($toString == null) {
+        $toString = "Option{"
+          + "__typename=" + __typename + ", "
+          + "allowFreeText=" + allowFreeText + ", "
+          + "id=" + id + ", "
+          + "label=" + label
+          + "}";
+      }
+      return $toString;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (o instanceof Option) {
+        Option that = (Option) o;
+        return this.__typename.equals(that.__typename)
+         && this.allowFreeText == that.allowFreeText
+         && this.id.equals(that.id)
+         && this.label.equals(that.label);
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      if (!$hashCodeMemoized) {
+        int h = 1;
+        h *= 1000003;
+        h ^= __typename.hashCode();
+        h *= 1000003;
+        h ^= Boolean.valueOf(allowFreeText).hashCode();
+        h *= 1000003;
+        h ^= id.hashCode();
+        h *= 1000003;
+        h ^= label.hashCode();
+        $hashCode = h;
+        $hashCodeMemoized = true;
+      }
+      return $hashCode;
+    }
+
+    public static final class Mapper implements ResponseFieldMapper<Option> {
+      @Override
+      public Option map(ResponseReader reader) {
+        final String __typename = reader.readString($responseFields[0]);
+        final boolean allowFreeText = reader.readBoolean($responseFields[1]);
+        final String id = reader.readString($responseFields[2]);
+        final String label = reader.readString($responseFields[3]);
+        return new Option(__typename, allowFreeText, id, label);
+      }
+    }
+  }
+}

